index/segment/mem/index.go

```diff
@@ -0,0 +1,225 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package mem
+
+import (
+	"crypto/md5"
+	"sync"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/segment"
+
+	"github.com/RoaringBitmap/roaring"
+	"github.com/uber-go/atomic"
+)
+
+// TODO(prateek): move to options
+const initialSize = 1024 * 1024
+
+// TODO(prateek): investigate impact of native heap
+type memIndex struct {
+	opts   Options
+	hashFn doc.HashFn
+
+	docIDGen *atomic.Uint32
+
+	// internal docID -> document
+	documentsLock sync.RWMutex
+	documents     map[segment.DocID]document
+
+	// fieldName -> (Term -> postingID)
+	fieldLock sync.RWMutex
+	fieldMap  map[string]termsDictionary
+
+	// postingID -> bitmap of docIDs
+	postingsList postingsList
+
+	// TODO(prateek): add a delete documents bitmap to optimise fetch
+}
+
+type document struct {
+	doc.Document
+	docID      segment.DocID
+	tombstoned bool
+}
+
+// New returns a new in-memory index.
+func New(opts Options) Segment {
+	hashFn := func(i doc.ID) doc.Hash {
+		return md5.Sum([]byte(i))
+	}
+
+	return &memIndex{
+		opts:   opts,
+		hashFn: hashFn,
+
+		docIDGen: atomic.NewUint32(0),
```

Comment:
Instead of making this per-segment, perhaps it would be preferable to make it a `uint64` that is per index? By making it per-index we won't have to merge the stored fields when we merge segments since we can store them separately from the segments themselves. Deleting documents should be cheaper as well since we would only need to store a bitmap of the deleted document IDs.

Reply:
The deleted documents benefit is fair, but consider that users don't have the internal document id when specifying a delete. We're going to have to handle delete compactions in addition to this anyway. Same thing for stored fields, regardless of internal doc id, we'd need to compact across the same external doc id. 

I have a few more concerns with this:
- it'd add contention to the write path (e.g. if we ran multiple mem segments to increase write throughput, they'd all compete for this resource)
- it'd increase the memory footprint of the bitmaps by a lot 
- `roaring` doesn't support uint64 bitmaps

Reply:
I don't see how the fact that users don't have the internal ID would factor in, we would need to look up the ID anyway, and so would already have access to it. We will need to handle compactions for stored fields, but we can do that independently of actual merges so we can optimize when it makes the most sense to do it. In regards to the other concerns:

- I don't see how we would have contention given that we will likely only have a single open segment at a time. Even if we did have multiple open segments, getting a new ID is just an atomic increment, and should be orders of magnitude less expensive than actually inserting documents into the index.
- If we apply delta encoding, the memory footprint shouldn't actually increase.
- We can coerce the `uint64` to `uint32` to start if we want to use Roaring until we use a more specialized approach which natively supports `uin64`.

Reply:
I don't follow. If we did Deletes/Updates by first retrieving the given document's internal id, and correspondingly deleting/updating the doc, then yeah, we'd get some benefit by having a doc id over the index lifecycle. 

But, we're going to do updates/deletes identical to insert. i.e. if we did and insert, and then a delete - we'd have 2 internal doc ids, each pointing to distinct documents that are identical until we compacted. The same thing for updates. i.e. we'd have to compact the fields for all documents. 

What am I missing?

Reply:
Might be easier to discuss in person and see where the ID per segment takes us to start. Like some of the other things, it's probably not worth getting hung up on this now lol.